sklearn/metrics/pairwise.py

```diff
@@ -256,6 +264,166 @@ def euclidean_distances(X, Y=None, Y_norm_squared=None, squared=False,
     return distances if squared else np.sqrt(distances, out=distances)
 
 
+# Pairwise distances in the presence of missing values
+def masked_euclidean_distances(X, Y=None, squared=False,
+                               missing_values="NaN", copy=True, **kwargs):
+    """
+    Considering the rows of X (and Y=X) as vectors, compute the
+    distance matrix between each pair of vectors. Similarly, if
+    Y is not X, then compute the distance matrix between each
+    pair of vectors (i.e., each row pair) in X and Y.
+
+    This function computes pairwise euclidean distance for vectors
+    in dense matrices X and Y with missing values in arbitrary
+    coordinates. The following formula is used for this:
+
+        dist(X, Y) = (X.shape[1] * 1 / ((dot(NX, NYT)))) *
```

Comment:
Hmm, I am not totally sure what you mean. I have edited the docstring to change it to what I think you meant -- please let me know what you think about the following: 

    This formulation zero-weights feature coordinates
    with missing value in either vector in the pair and up-weights the
    remaining coordinates. For instance, say we have two sample points (x1,
    y1) and (x2, NaN). To calculate the euclidean distance, first the square
    "distance" is calculated based only on the first feature coordinate, 
    as the second coordinate is missing in one of the samples,
    i.e., we get (x2-x1)**2. This squared distance is scaled-up by the ratio 
    of total number of coordinates to the number of available coordinates, 
    which in this case is 2/1 = 2. Now, we are left with 2*((x2-x1)**2). 
    Finally, if squared=False then the square root of this is evaluated 
    otherwise the value is returned as is.